### PR TITLE
fix(clerk-js): Trim emails, usernames and identifiers

### PR DESCRIPTION
**Pull request by desiprisg**
## Description

Trim email addresses, usernames and the text identifier as whitespace at the start and end of the value is invalid.

To achieve the trimming for the identifier, a `transformer` prop has been added to `useFormControl` that can be used for other purposes as well.

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

**Comment by nikosdouvlis** [comment]
@desiprisg please add a description and a changeset as this does change the behavior of the components

**Comment by desiprisg** [comment]
!preview

File: packages/clerk-js/src/ui/utils/useFormControl.ts [hunk 1]
```diff
@@ -80,30 +81,31 @@ export const useFormControl = <Id extends string>(
...
-  opts = opts || {
+  const options = opts || {
     type: 'text',
     label: '',
     isRequired: false,
     placeholder: '',
     options: [],
     defaultChecked: false,
   };
+  const transformer: NonNullable<typeof options.transformer> = options.transformer ? options.transformer : v => v;
```
Comment (panteliselef) [inline packages/clerk-js/src/ui/utils/useFormControl.ts:92]: 🙃
```suggestion
  const transformer: NonNullable<typeof options.transformer> = options.transformer ?? v => v;
```

File: packages/clerk-js/src/ui/primitives/Input.tsx [hunk 2]
```diff
@@ -76,12 +76,12 @@ export const Input = React.forwardRef<HTMLInputElement, InputProps>((props, ref)
...
-      ? {
-          pattern: '^.*@[a-zA-Z0-9\\-]+\\.[a-zA-Z0-9\\-\\.]+$',
-        }
-      : { type };
+  let value = rest.value;
+  let typeProps: Record<string, unknown> = { type };
+  if (typeof value === 'string' && type === 'email') {
+    value = value.trim();
+    typeProps = { pattern: '^.*@[a-zA-Z0-9\\-]+\\.[a-zA-Z0-9\\-\\.]+$' };
+  }
```
Comment (panteliselef) [inline packages/clerk-js/src/ui/primitives/Input.tsx:84]: ❓ 
1. Is it possible to have `type='email'` but not a value that is a string ?
2. I'd expect with the changes made in our `useFormControl` hook (below) that we wouldn't need to call  `value = value.trim();` manually here. Am i missing something ?

Comment (desiprisg) [reply packages/clerk-js/src/ui/primitives/Input.tsx:84]: I just followed typescript here

Comment (LekoArts) [reply packages/clerk-js/src/ui/primitives/Input.tsx:84]: +1 on the second question. Shouldn't there be a `transformer` in `packages/clerk-js/src/ui/components/SignUp/SignUpStart.tsx` for `emailAddress`?

File: packages/clerk-js/src/ui/utils/useFormControl.ts [hunk 3]
```diff
@@ -75,35 +78,51 @@ export type FormControlState<Id = string> = FieldStateProps<Id> & {
...
     type: 'info',
   });
 
   const onChange: FormControlState['onChange'] = event => {
-    if (opts?.type === 'checkbox') {
+    if (options?.type === 'checkbox') {
       return setCheckedInternal(event.target.checked);
     }
-    return setValueInternal(event.target.value || '');
+    return setValueInternal(applyTransformers(event.target.value || '', transformers));
```
Comment (panteliselef) [inline packages/clerk-js/src/ui/utils/useFormControl.ts:125]: 💯

**Review by panteliselef — APPROVED** [review]

File: packages/clerk-js/src/ui/primitives/Input.tsx (hunk 2 above)
Comment (desiprisg) [reply packages/clerk-js/src/ui/primitives/Input.tsx]: There is a default transformer applied for type email in `useFormControl`

Comment (LekoArts) [reply packages/clerk-js/src/ui/primitives/Input.tsx]: Yeah, looks good now

**Pull request merged**